renku/core/management/workflow/value_resolution.py

```diff
@@ -36,70 +38,80 @@ def apply_run_values(workflow: Union[CompositePlan, Plan], values: Dict[str, Any
     - Value propagated to a parameter from the source of a ParameterLink
     """
 
-    if isinstance(workflow, Plan):
-        return apply_single_run_values(workflow, values)
-
-    return apply_composite_run_values(workflow, values)
-
+    def __init__(self, plan: AbstractPlan, values: Dict[str, Any]):
+        self._values = values
+        self.missing_parameters: Set[str] = {}
+        self._plan = plan
 
-def apply_single_run_values(workflow: Plan, values: Dict[str, Any] = None) -> None:
-    """Applies values and default_values to a workflow."""
-    if not values:
-        return workflow
+    def apply(self) -> AbstractPlan:
+        """Applies values and default_values to a potentially nested workflow.
 
-    for param in workflow.inputs + workflow.outputs + workflow.parameters:
-        if param.name in values:
-            param.actual_value = values[param.name]
+        :returns: The ``AbstractPlan`` with the user provided values set.
+        """
+        self._apply_single_run_values() if isinstance(self._plan, Plan) else self._apply_composite_run_values()
+        return self._plan
 
-    return workflow
+    def _apply_single_run_values(self) -> None:
+        """Applies values and default_values to a workflow."""
+        if not self._values:
+            return
 
+        values_keys = set(self._values.keys())
+        for param in chain(self._plan.inputs, self._plan.outputs, self._plan.parameters):
+            if param.name in self._values:
+                param.actual_value = self._values[param.name]
+                values_keys.discard(param.name)
 
-def apply_composite_run_values(workflow: CompositePlan, values: Dict[str, Any] = None) -> None:
-    """Applies values and default_values to a nested workflow."""
+        self.missing_parameters = values_keys
 
-    if values:
-        if "parameters" in values:
-            # NOTE: Set mapping parameter values
-            apply_parameters_values(workflow, values["parameters"])
+    def _apply_composite_run_values(self) -> None:
+        """Applies values and default_values to a nested workflow."""
 
-        if "steps" in values:
-            for name, step in values["steps"].items():
-                child_workflow = next((w for w in workflow.plans if w.name == name), None)
-                if not child_workflow:
-                    raise errors.ChildWorkflowNotFound(name, workflow.name)
+        if self._values:
+            if "parameters" in self._values:
+                # NOTE: Set mapping parameter values
+                self._apply_parameters_values()
 
-                apply_run_values(child_workflow, step)
+            if "steps" in self._values:
+                for name, step in self._values["steps"].items():
+                    child_workflow = next((w for w in self._plan.plans if w.name == name), None)
+                    if not child_workflow:
+                        self.missing_parameters.add(f"steps.{name}")
```

Comment:
Maybe we should differentiate between a child workflow not existing and a parameter on a child workflow not existing